File: Object oriented programming (II sem)/OOP in cpp/Animal.cpp
#include"Animal.h"

Animal::Animal(string name, int strength, int initiative, char icon, Position pos, World& world)
	: Organism(name, strength, initiative, icon, pos, world) {
}
void Animal::action() {
	Position newPos = pickRandomNeighbourField(STANDARD_RANGE);
	if (world.isEmpty(newPos)) {
		move(newPos);
	}
	else {
		collision(world.getBoard()[newPos.x][newPos.y]);
	}
}
void Animal::collision(Organism* other) {
	Position posToMove(other->getPos());
	if (world.getBoard()[posToMove.x][posToMove.y]->getIcon() == icon) {			//there is the same species on the field
		reproduce();
	}
	else {																					//ther is different species on the field
		//if(other->getIcon() == )
		if (other->getIcon() == GUARANA_ICON) strength += 3;
		if (other->getIcon() == TURTLE_ICON && strength < 5) return;
		if (other->getIcon() == ANTYLOPE_ICON && world.chanceOfBeeingTrue(0.5f)) {
			Position emptyNeighbourField = other->findEmptyNeighbourField(other->getPos());
			if (emptyNeighbourField.isCorrect()) {
				other->move(emptyNeighbourField);
				move(posToMove);
			}			
		}
		if (strength >= other->getStrength()) {
			Announcer::addAnnounce(name + " ate " + other->getName());
			world.killOrganism(other);
			move(posToMove);
		}
		else {
			Announcer::addAnnounce(other->getName() + " killed " + name);
			world.killOrganism(this);
		}
	}
}
void Animal::move(Position pos) {
	position = pos;
}
void Animal::reproduce() {
	Position newAnimalPos = findEmptyNeighbourField(position);
	if (!newAnimalPos.isCorrect()) return;										//no place to put new animal
	Announcer::addAnnounce(name + " got reporduced");
	world.createNewOrganism(icon, newAnimalPos);
}
bool Animal::isStronger(Organism* other) {
	if (this->getStrength() > other->getStrength()) return true;
	else return false;
}
Animal::~Animal() {
}

File: Object oriented programming (II sem)/OOP in java/Animal.java
import java.awt.*;

public class Animal extends Organism {

    public Animal(String name, int strength, int initiative, Color color, Position position, World world) {
        super(name, strength, initiative, color, position, world);
    }

    @Override
    public void action() {
        Position desiredPosition = getDesiredPosition(1);
        if (world.isEmpty(desiredPosition)) {
            move(desiredPosition);
        } else {
            collision(world.getOrganism(desiredPosition));
        }
        updateAge();
    }

    @Override
    public void move(Position position) {
        this.position = position;
        world.addAnnounce(announcer.move(this, position));
    }

    @Override
    public void collision(Organism other) {
        if (this.getClass() == other.getClass()) {
            reproduce();
        } else {
            fight(other);
        }
    }

    public void reproduce() {
        if (!isSurroundedByItsSpecies()) {
            Position newOrganismPosition;
            do {
                newOrganismPosition = getDesiredPosition(1);
                if (world.isEmpty(newOrganismPosition)) break;
            } while (world.getOrganism(newOrganismPosition).getClass() != this.getClass());
            world.newLivingOrganism(this, newOrganismPosition);
            world.addAnnounce(announcer.reproduction(this));
        }
    }

    public void fight(Organism other) {
        Position desiredPosition = other.getPosition();
        if (other.blockedAttack(this)) return;
        if (other.runAway()) {
            this.move(desiredPosition);
            return;
        }
        if (this.isStrongerOrEqualThen(other)) {
            world.addAnnounce(announcer.death(other, this));
            if (other.gaveStrengthIncrease()) this.increaseStrength(3);
            other.setPositionToNull();
            move(desiredPosition);
        } else {
            world.addAnnounce(announcer.death(this, other));
            this.setPositionToNull();
        }
    }
}
